easybuild/tools/containers.py

```diff
@@ -0,0 +1,305 @@
+CONT_TYPE_SINGULARITY = 'singularity'
+
+DOCKER = 'docker'
+LOCALIMAGE = 'localimage'
+SHUB = 'shub'
+SINGULARITY_BOOTSTRAP_TYPES = [DOCKER, LOCALIMAGE, SHUB]
+
+
+SINGULARITY_TEMPLATE = """
+Bootstrap: %(bootstrap)s
+From: %(from)s
+
+%%post
+%(install_os_deps)s
+
+# upgrade easybuild package automatically to latest version
+pip install -U easybuild
+
+# change to 'easybuild' user
+su - easybuild
+
+eb %(easyconfigs)s --robot --installpath=/app/ --prefix=/scratch --tmpdir=/scratch/tmp
+
+# exit from 'easybuild' user
+exit
+
+# cleanup
+rm -rf /scratch/tmp/* /scratch/build /scratch/sources /scratch/ebfiles_repo
+
+%%runscript
+eval "$@"
+
+%%environment
+source /etc/profile
+module use /app/modules/all
+module load %(mod_names)s
+
+%%labels
+
+"""
+
+
+_log = fancylogger.getLogger('tools.containers')  # pylint: disable=C0103
+
+
+def parse_container_base(base):
+    """Parse value passed to --container-base option."""
+    if base:
+        base_specs = base.split(':')
+        if len(base_specs) > 3 or len(base_specs) <= 1:
+            error_msg = '\n'.join([
+                "Invalid format for --container-base, must be one of the following:",
+                '',
+                "--container-base localimage:/path/to/image",
+                "--container-base shub:<image>:<tag>",
+                "--container-base docker:<image>:<tag>",
+            ])
+            raise EasyBuildError(error_msg)
+    else:
+        raise EasyBuildError("--container-base must be specified")
+
+    # first argument to --container-base is the Singularity bootstrap agent (localimage, shub, docker)
+    bootstrap_agent = base_specs[0]
+
+    # check bootstrap type value and ensure it is localimage, shub, docker
+    if bootstrap_agent not in SINGULARITY_BOOTSTRAP_TYPES:
+        known_bootstrap_agents = ', '.join(SINGULARITY_BOOTSTRAP_TYPES)
+        raise EasyBuildError("Bootstrap agent in container base spec must be one of: %s" % known_bootstrap_agents)
+
+    res = {'bootstrap_agent': bootstrap_agent}
+
+    for idx, base_spec in enumerate(base_specs[1:]):
+        res.update({'arg%d' % (idx + 1): base_spec})
+
+    return res
+
+
+def generate_singularity_recipe(easyconfigs, container_base):
+    """Main function to Singularity definition file and image."""
+
+    cont_path = container_path()
+
+    # make sure location to write container recipes & images exists
+    mkdir(cont_path, parents=True)
+
+    base_specs = parse_container_base(container_base)
+
+    # extracting application name,version, version suffix, toolchain name, toolchain version from
+    # easyconfig class
+
+    bootstrap_agent = base_specs['bootstrap_agent']
+
+    base_image, base_image_tag = None, None
+
+    # with localimage it only takes 2 arguments. --container-base localimage:/path/to/image
+    # checking if path to image is valid and verify image extension is '.img' or '.simg'
+    if base_specs['bootstrap_agent'] == LOCALIMAGE:
+        base_image = base_specs['arg1']
```

Comment:
@boegel 

Running into some weird issues with localimage 

for instance if you have a space character between ":" and the container path I get the following

```
[siddis14@amrndhl1157 easybuild-framework]$ eb Anaconda3-4.2.0.eb -C --container-base localimage: ~/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
== temporary log file in case of crash /tmp/eb-iXLynP/easybuild-ScZ5Ui.log
ERROR: Singularity base image at specified path does not exist:
```

Shell expansion does not work

```
[siddis14@amrndhl1157 easybuild-framework]$ eb Anaconda3-4.2.0.eb -C --container-base localimage:~/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
== temporary log file in case of crash /tmp/eb-MrrU23/easybuild-ht00Iw.log
ERROR: Singularity base image at specified path does not exist: ~/.local/easybuild/containers/git-lfs-1.1.1/
```

This works

```
[siddis14@amrndhl1157 easybuild-framework]$ eb Anaconda3-4.2.0.eb -C --container-base localimage:$HOME/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
== temporary log file in case of crash /tmp/eb-b2rOE1/easybuild-pqXfed.log
ERROR: Invalid image extension '' must be .img or .simg
```



I think we can trim any space and newline characters before processing the argument. I think this should be done in parse_container_base method 

If you can get the shell expansion working that would be nice if not we can create an issue to track this for future.

Reply:
This is not so easy to fix... If you use a space, `eb` will take `localimage` as an argument to `--container-base`, and the specified path of the image as if it were an easyconfig file, and it will no longer be related to `--container-base` at all...
All arguments on the `eb` command that are not tied to an option are assumed to be easyconfig files.

So, this is not so trivial to fix, unless we introduce another option like `--container-base-args` or something like that:

```
eb Anaconda3-4.2.0.eb -C --container-base localimage --container-base-args $HOME/.local/easybuild/containers/git-lfs-1.1.1/ --experimental
```

Reply:
@boegel no worries let's hold off on this if it is too much work.

Is the shell expansion something that can be fixed easy so if user does the following ``--container-base localimage:~/some.img`` it will resolve properly to the $HOME directory instead of throwing an error.

Reply:
The tab completion or shell expansion of ~ won't work because of the preceding `:`, we can't easily fix that unless we go with a separate option like `--container-base-args`

Reply:
okay let's leave it for now and remember this for future.